Refactor and optimize uslope

tinecolman opened the pull request:
I've gather the formulas for the different slope limiters into a module called slope_types. For each slope type, there is a pure function which the compiler should inline thanks to the additional compiler flags in the Makefile.
This structure reduces code duplication and makes it easier to optimize the existing slope limiters.

I've rewritten minmod/average for avoid the division and casting from int to float in the inner loop. This also makes the calculation of dlft and drgt the same as for the other regular slope types.

The slope module also contains a subroutine for each slope_type which is doing the calculation for nvector elements. Most of these call their corresponding slope_type function for each dimension.
In these subroutines, the center value is now prefetched into a scalar, and reused for the calculation of dlft and drgt, reducing the number of load operations. This gives a small performance gain.

The main subroutine uslope is now effectively a wrapper to call the requested calc_uslope subroutine.
I've rearranged the parts for different NDIM to have less duplication. For HYDRO slope_type 1, the structure is a bit strange due to their being a different formulation for 3D for unclear reasons. Since I didn't want to change the functionality, I've kept both.
 
For MHD, the calculation of the B-field slopes are now gathered at the end of the routine, to make it easier to merge hydro and mhd files in the future.

tinecolman committed:
splits uslope in 1d, 2d and 3d version

tinecolman committed:
create slope_types module and add 3d minmod slope to it

tinecolman committed:
add 3d vanLeer and moncen slopes to slope_types module

tinecolman committed:
add generalized slope_type==8 to slope_types module

tinecolman committed:
some cleanup for uslope3d

tinecolman committed:
split if for slope_type 1 and 2 in 2D case

tinecolman committed:
fixes

tinecolman committed:
cleanup mhd uslope

tinecolman committed:
in uslope, retrieve values of cell center and neighbors

tinecolman committed:
uslope: gather cell values first

nbrucy committed:
set npartmax to 0 if pic is false (#150)

This fixes #149.

Sod tube test modified to check that combinaison of parameters.

---------

Co-authored-by: Noe Brucy <noe.brucy@ens-lyon.fr>

tinecolman committed:
Fix nlevelmax_sink not working as intended (#151)

Patrick Hennebelle reported that sinks were not forming properly when
using this. He traced it to a problem with clump_mass4. I looked further
into it and found places where nlevelmax should be replaced by
nlevelmax_sink.

I also fixed the issue were nlevelmax_sink could not be combined with
sink_refine.

I updated the sink_spawning test namelist to catch this issue.
I verified there are no issues for sink feedback, by adding
nlevelmax_sink to the tests that cover these.

tinecolman committed:
Merge branch 'slope_module' into optim_uslope

tinecolman committed:
combine 1D, 2D and 3D blocks in uslope

tinecolman committed:
remove gather_local_values again, keep fetch of center value

tinecolman committed:
add other version of minmod for 1D and 2D

tinecolman committed:
follow more closely original structure

tinecolman committed:
sync mhd

tinecolman committed:
fetch center value optimization for b field slope

tinecolman committed:
add compiler flag for in-lining

tinecolman committed:
fix

tinecolman committed:
add flag for intel

tinecolman commented:
An issue is that not all of the relevant slope_type - NDIM combinations are currently covered by the test suite.

tinecolman committed:
typo

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nbrucy commented:
> An issue is that not all of the relevant slope_type - NDIM combinations are currently covered by the test suite.

So maybe we need to finish #141 before merging this one?

tinecolman commented:
<img src="https://github.com/user-attachments/assets/20c71ae2-3013-42e7-8e3b-80b174ed4055" width=60%>

I get a speedup of 5% on the uniform grid sedov benchmark with this.

tinecolman commented:
Something is not working for intel compiler. Currently investigating.

tinecolman committed:
force inlining for intel

tinecolman commented:
For intel, the performance is the same as before: no gain but also no loss.

tinecolman committed:
Merge branch 'dev' into optim_uslope

tinecolman committed:
Merge branch 'dev' into optim_uslope

tinecolman committed:
fix to slope_type 4

tinecolman commented:
I fixed a bug for slope_type=4, which was revealed by the sod parameter sweep.
Tests confirm that now everything works as before for the case HYDRO 1D.

tinecolman committed:
no B-field slope for NDIM=1

tinecolman committed:
fix slope_type 3 in 2D

tinecolman commented:
Using the orzag-tang test, I manually confirmed that the new implementation is now working as before for 2D MHD, with all slope_type options (1,2,3).

tinecolman commented:
Using the collapse-baro test, I manually confirmed that the new implementation is working as before for 3D MHD, with all slope_type options (1,2,3,7,8).

tinecolman commented:
Using the implosion, mixing, cooling-frig and sedov3d tests, I manually confirmed that the new implementation is now working as before for 2D and 3D HYDRO with all slope_type options (1,2,3,7,8).

tinecolman committed:
create calc_uslope subroutines for each slope_type



I've moved the if-check inwards, so that the outer loop i,j,k is now only done once. Then I created vectorized subroutines for each slope_type, so that uslope is now effectively a wrapper routine to select the requested slope type.
By keeping the dimensions of input and output arrays the same as before, the compiler properly inlines and vectorizes. And so the execution time is unaffected.

This structure will allow to move the call to uslope inside trace, bypassing a slow store-load operation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

tinecolman commented:
I've moved the if-check inwards, so that the outer loop i,j,k is now only done once. Then I created vectorized subroutines for each slope_type, so that uslope is now effectively a wrapper routine to select the requested slope type.
By keeping the dimensions of input and output arrays the same as before, the compiler properly inlines and vectorizes. And so the execution time is unaffected.

This structure will allow to move the call to uslope inside trace, bypassing a slow store-load operation. This change will be made in a separate PR once this one, and the refactoring of trace is accepted

tinecolman committed:
Merge branch 'dev' into optim_uslope

rteyssier committed:
Merge branch 'dev' into optim_uslope

rteyssier committed:
Merge branch 'dev' into optim_uslope

rteyssier commented:
It looks like we have some conflicts here...

tinecolman committed:
Merge branch 'dev' into optim_uslope

tinecolman commented:
> It looks like we have some conflicts here...

It's because I deleted mhd/hydro_parameters.90 in the other PR. Fixed it!
